File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res, sol = [], []
        n = len(candidates)

        def backtrack(i, summ):
            if i > n or summ > target:
                return 
            
            if summ == target:
                temp = sol[:]
                temp.sort()
                if temp not in res: res.append(temp)
                return 

            for j in range(i, n):
                sol.append(candidates[j])
                backtrack(j + 1, summ + candidates[j])
                sol.pop()

        backtrack(0, 0)
        return res

File: Data Structures & Algorithms/combinations/submission-2.go
func combine(n int, k int) [][]int {
 	sol := []int {} 
	res := [][]int {}

	var backtrack func(i int) 

	backtrack = func(i int) {
		if len(sol) == k {
			res = append(res, append([]int{}, sol...))
			return 
		}

		for j := i; j < n + 1; j++ {
			sol = append(sol, j)
			backtrack(j + 1)
			sol = sol[:len(sol) - 1]
		}
	}

	backtrack(1)
	return res
}
